Sourcegraph 3.16 release blog post (#951)
sourcegraph/about: blogposts/2020/announcing-sourcegraph-3.16.md

Sourcegraph 3.16 includes major improvements to syntax highlighting throughout Sourcegraph. In addition to general improvements, six new languages are supported (Smarty, Ethereum / Solidity / Vyper, Cuda, COBOL, vb.NET, and ASP.NET), and 30 new file extensions are now detected.

Additionally, this release adds best-in-class syntax highlighting to code diffs. This richer view of diffs can easily be accessed from your pull request (or merge request) using the Sourcegraph browser extension by clicking on the Sourcegraph button in a file header. Alternatively, you can compare revisions directly from Sourcegraph by going to the compare section from a repository page. You can also see this in action on your [campaigns](#campaigns) to make changes across all the repositories at your organization.

For an example of syntax highlighting on a diff, check out [a recent change in the Sourcegraph repository](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/commit/3fa28d025d615894867f550dfdbf5fb5b6de958f). Make sure to try it out in both dark and light mode!

## Easier to read queries in plain text mode

> TODO: video
> 
> <!--
> <p class="container">
>   <div style="padding:56.25% 0 0 0;position:relative;">
>     <iframe src="https://www.youtube.com/embed/{VIDEO_ID}" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
>   </div>
>   <p style="text-align: center"><a href="https://youtu.be/{VIDEO_ID}" target="_blank">View on YouTube</a></p>
> </p>
> -->

Reviewer: ```suggestion
<p class="container">
  <div style="padding:56.25% 0 0 0;position:relative;">
    <iframe src="https://www.youtube.com/embed/{ZtnqWhIcLeg}" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
  </div>
  <p style="text-align: center"><a href="https://youtu.be/{ZtnqWhIcLeg}" target="_blank">View on YouTube</a></p>
</p>
```

@christinaforney added video.

- If the experimental feature flag `codeInsights` is enabled, extensions can contribute content to directory pages through the experimental `ViewProvider` API. [#10236](https://github.com/sourcegraph/sourcegraph/pull/10236)
  - Directory pages are then represented as an experimental `DirectoryViewer` in the `visibleViewComponents` of the extension API. **Note: This may break extensions that were assuming `visibleViewComponents` were always `CodeEditor`s and did not check the `type` property.** Extensions checking the `type` property will continue to work. [#10236](https://github.com/sourcegraph/sourcegraph/pull/10236)
- [Major syntax highlighting improvements](https://github.com/sourcegraph/syntect_server/pull/29), including:
  - 228 commits / 1 year of improvements to the syntax highlighter library Sourcegraph uses ([syntect](https://github.com/trishume/syntect)).
  - 432 commits / 1 year of improvements to the base syntax definitions for ~36 languages Sourcegraph uses ([sublimehq/Packages](https://github.com/sublimehq/Packages)).
  - 30 new file extensions/names now detected.
  - Likely fixes other major instability and language support issues. #9557

>   - Added [Smarty](#2885), [Ethereum / Solidity / Vyper)](#2440), [Cuda](#5907), [COBOL](#10154), [vb.NET](#4901), and [ASP.NET](#4262) syntax highlighting.

Author: Nope - good catch. Fixed.

The [changelog for this and previous releases](https://github.com/sourcegraph/sourcegraph/blob/master/CHANGELOG.md#3.16) is available on GitHub.

## Thank you

Thank you to the many people who contributed to Sourcegraph since the last release!

- [@blanet](https://github.com/blanet)
- [@dobrou](https://github.com/dobrou)
- [@yg](https://github.com/yg)
- [@souldzin](https://github.com/souldzin)
- [@YouveGotMeowxy](https://github.com/YouveGotMeowxy)
- [@b-j-p](https://github.com/b-j-p)
- [@jlangston](https://github.com/jlangston)
- [@yaohui-wyh](https://github.com/yaohui-wyh)
- [@eight04](https://github.com/eight04)
- [@marco-c](https://github.com/marco-c)
- [@jhutchings1](https://github.com/jhutchings1)
- [@philjones](https://github.com/philjones)

**Deploy or upgrade:** [Local](https://docs.sourcegraph.com/#quickstart-guide) | [AWS](https://github.com/sourcegraph/deploy-sourcegraph-aws) | [DigitalOcean](https://marketplace.digitalocean.com/apps/sourcegraph?action=deploy&refcode=48dfb3ccb51c) | [Kubernetes cluster](https://github.com/sourcegraph/deploy-sourcegraph)

> From the entire Sourcegraph team ([@srcgraph](https://twitter.com/srcgraph)), happy coding!

Reviewer 2: ```suggestion
## Share your feedback

We want to hear what you think about Sourcegraph and the new updates in 3.16. Post on Twitter (remember to include ([@srcgraph](https://twitter.com/srcgraph)) and we'll send you Sourcegraph swag!

Any questions about Sourcegraph? Get in touch on Twitter <a href="https://twitter.com/srcgraph">@srcgraph</a>, file an issue in our{' '}<a href="https://github.com/sourcegraph/sourcegraph/issues">public issue tracker</a>, or email{' '}<a href="mailto:feedback@sourcegraph.com">feedback@sourcegraph.com</a>. We look forward to hearing from you!

From the [entire Sourcegraph team](https://about.sourcegraph.com/company/team)), happy coding!
```

---
title: "Sourcegraph 3.16: Version contexts, campaigns, and major syntax highlighting improvements"
author: Christina Forney
publishDate: 2020-05-20T10:00-07:00
tags: [
  blog
]
slug: sourcegraph-3.16
heroImage: https://about.sourcegraph.com/sourcegraph-mark.png
published: true
---

Explore, navigate, and better understand all code, everywhere, faster with Sourcegraph Universal Code Search. Uber, Lyft, Yelp, and other enterprises rely on Sourcegraph to improve developer productivity and manage large-scale code changes.

<div style="padding-left: 2rem">

[**📌 Introducing version contexts to search past releases**](#introducing-version-contexts-to-search-past-releases)<br />

[**📣 Smoother getting started experience for campaigns**](#smoother-getting-started-experience-for-campaigns)<br />

> [**🎨 Best in class syntax highlighting**](#best-in-class-syntax-highlighting)<br />

Reviewer 3: "Best-in-class" - add hyphens because that is how I usually see it, and it also makes it easier to parse for non-native English speakers.

```json
"experimentalFeatures": [
  "versionContexts": [
    {
      "name": "srcgraph 3.16",
      "revisions": [
        {
          "repo": "github.com/sourcegraph/sourcegraph",
          "rev": "v3.16.0"
        },
        {
          "repo": "github.com/sourcegraph/zoekt",
          "rev": "b56036a3b745033badc48807bd67e91cd8d73bdf"
        }
      ]
    }
  ]
]
```

## Smoother getting started experience for campaigns

<p class="container">
  <div style="padding:56.25% 0 0 0;position:relative;">
    <iframe src="https://www.youtube.com/embed/aqcCrqRB17w" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
  </div>
  <p style="text-align: center"><a href="https://youtu.be/aqcCrqRB17w" target="_blank">View on YouTube</a></p>
</p>

Besides multiple smaller bug fixes and performance improvements, we aimed to make it easier and more straightforward to start using campaigns on your own Sourcegraph instance by providing clearer instructions in the onboarding process.

> We also made it easier to setup the [campaigns webhooks](https://docs.sourcegraph.com/user/campaigns/getting_started), which we highly recommend to reduce load on the code host and to improve the speed with which campaigns are updated in Sourcegraph.

Reviewer 3: ```suggestion
We also made it easier to set up [campaigns webhooks](https://docs.sourcegraph.com/user/campaigns/getting_started), which we highly recommend to reduce load on the code host and to improve the speed with which campaigns are updated in Sourcegraph.
```

The [documentation for campaigns](https://docs.sourcegraph.com/user/campaigns) has also undergone a complete overhaul and now features a video that explains what campaigns can do and is included above.

> Campaign changeset usage counts including changesets created, added, and merged will be sent back in pings.

Reviewer 3: ```suggestion
```

  </div>
  <p style="text-align: center"><a href="https://youtu.be/aqcCrqRB17w" target="_blank">View on YouTube</a></p>
</p>

Campaign changeset usage counts including changesets created, added, and merged will be sent back in pings. [#10591](https://github.com/sourcegraph/sourcegraph/pull/10591)

## Best in class syntax highlighting

TODO: Erik is creating screenshots

We have made major improvements to syntax highlighting throughout Sourcegraph. In addition to general improvements, six new languages are supported (Smarty, Ethereum / Solidity / Vyper, Cuda, COBOL, vb.NET, and ASP.NET), and 30 new file extensions are now detected. 

Additionally, this release adds best-in-class syntax highlighting to code diffs. This richer view of diffs can easily be accessed from your pull request (or merge request) using the Sourcegraph browser extension by clicking on the Sourcegraph button in a file header. Alternatively, you can compare revisions directly from Sourcegraph by going to the compare section from a repository page. You can also see this in action on your [campaigns](#campaigns) to make changes across all the repositories at your organization. 

> Take a look at the syntax highlighting on a diff that &lt;WHAT IS THIS DIFF ABOUT> in the Sourcegraph repository.

Author: @eseliger - can you find a link to nice diff that shows off syntax highlighting to include here?

Reviewer 4: The link to the above screenshots is https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/commit/3fa28d025d615894867f550dfdbf5fb5b6de958f

Reviewer 5: Added GIF that transitions between light and dark mode Sourcegraph diffs

## Performance improvements for precise code intelligence

![performance-improvements-code-intel](/blog/3-16-ts-to-go-lsif.png "performance-improvements-code-intel")

As of Sourcegraph 3.16, our precise code intelligence backend has been rewritten from TypeScript to Go. This effort is part of a larger effort to aggressively optimize conversion and querying of LSIF data, which will start becoming more apparent in upcoming releases. Even with this initial improvement, we’re seeing improvements on precise code intelligence query performance. The following graph shows the five-run average duration of our [integration test](https://github.com/sourcegraph/sourcegraph/tree/5f51043ad2130a1acdcfca8b969f907cd03a220d/internal/cmd/precise-code-intel-test) before and after the rewrite.

## Discover useful scopes with `repogroup` autocompletion

> TODO: video

Reviewer: ```suggestion
<img src="https://storage.googleapis.com/sourcegraph-assets/repogroup_completion.gif" alt="Repogroup autocompletion">
```

@christinaforney this is a simple change, so I think a video with voiceover would be overkill here. I put together a gif instead.

Author: Nice! Thank you!

[**🔑 Check syncing status of repository and user permissions**](#check-syncing-status-of-repository-and-user-permissions)<br />

[**📊 Understanding user activity of your instance**](#understanding-user-activity-of-your-instance)<br />

[**ⓘ Improvements to basic code intelligence for several languages**](#improvements-to-basic-code-intelligence-for-several-languages)<br />

[**📝 Changelog**](#changelog)<br />
Every detail that changed in this release

[**🎖️ Thank you**](#thank-you)<br />
Sourcegraph couldn't be what it is without the community.

</div>

**Deploy or upgrade:** [Local](https://docs.sourcegraph.com/#quickstart-guide) | [AWS](https://github.com/sourcegraph/deploy-sourcegraph-aws) | [DigitalOcean](https://marketplace.digitalocean.com/apps/sourcegraph?action=deploy&refcode=48dfb3ccb51c) | [Kubernetes cluster](https://github.com/sourcegraph/deploy-sourcegraph)

<div class="alert alert-info">
  <p>
    <strong>LEARN MORE:</strong><br />
    <a href="https://info.sourcegraph.com/online-meetup-may21?utm_source=blog">
      Online meetup: Sourcegraph 3.16 - Thursday, May 21 at 10AM PDT
    </a>
  </p>
</div>

## Introducing version contexts to search past releases

> TODO: video

Author: @attfarhan to add video link

In this release, any non-master branches that are specified in the context will use our non-indexed search. This means that results may be slow for larger deployments. We are working on adding support for indexing additional branches that are specified within the context, and will include this in a future release.

To take advantage of of this feature, add a `versionContexts` block to the experimental features in your site configuration:

```json
"experimentalFeatures": [
  "versionContexts": [
    {
      "name": "srcgraph 3.16",
      "revisions": [
        {
          "repo": "github.com/sourcegraph/sourcegraph",
          "ref": "v3.16.0"
        },
        {
          "repo": "github.com/sourcegraph/zoekt",
          "ref": "b56036a3b745033badc48807bd67e91cd8d73bdf"
        }
      ]
    }
  ]
]
```

> ## Campaigns

Author: @sqs or @mrnugget to write initial draft (I am happy to help edit!).

Reviewer 6: ```suggestion
## Campaigns

Besides multiple smaller bug fixes and performance improvements, we aimed to make it easier and more straightforward to start using campaigns on your own Sourcegraph instance by providing clearer instructions in the onboarding process.

We also made it easier to setup the [campaigns webhooks](https://docs.sourcegraph.com/user/campaigns/getting_started), which we highly recommend to reduce load on the code host and to improve the speed with which campaigns are updated in Sourcegraph.

The [documentation for campaigns](https://docs.sourcegraph.com/user/campaigns) has also undergone a complete overhaul and now features a video that explains what campaigns can do:
```

<p class="container">
  <div style="padding:56.25% 0 0 0;position:relative;">
    <iframe src="https://www.youtube.com/embed/aqcCrqRB17w" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
  </div>
  <p style="text-align: center"><a href="https://youtu.be/aqcCrqRB17w" target="_blank">View on YouTube</a></p>
</p>

Campaign changeset usage counts including changesets created, added, and merged will be sent back in pings. [#10591](https://github.com/sourcegraph/sourcegraph/pull/10591)

## Best in class syntax highlighting

> TODO: Erik is creating screenshots

Author: @eseliger to add screenshots for syntax highlighting

Reviewer 4: <img width="1147" alt="SourcegraphDark" src="https://user-images.githubusercontent.com/19534377/82244479-9a5e0b80-9941-11ea-838d-ad90ec506674.png">
<img width="1147" alt="SourcegraphLight" src="https://user-images.githubusercontent.com/19534377/82244493-9d58fc00-9941-11ea-936f-3483cee8aabf.png">
<img width="1022" alt="GitHub" src="https://user-images.githubusercontent.com/19534377/82244495-9e8a2900-9941-11ea-9b4f-f0c07c1e4d4d.png">

do you want me to upload them somewhere else? I added the same diff on our end both light and dark (I don't know any other major codehost that supports that!)
and GitHubs highlighting (which is really plain in this particular case - but there are examples where it isn't _that_ plain at least. But overall I think our light theme is much more diverse in colors and accounts much better for the semantic meaning of the tokens highlighted.

We have made major improvements to syntax highlighting throughout Sourcegraph. In addition to general improvements, six new languages are supported (Smarty, Ethereum / Solidity / Vyper, Cuda, COBOL, vb.NET, and ASP.NET), and 30 new file extensions are now detected. 

Additionally, this release adds best-in-class syntax highlighting to code diffs. This richer view of diffs can easily be accessed from your pull request (or merge request) using the Sourcegraph browser extension by clicking on the Sourcegraph button in a file header. Alternatively, you can compare revisions directly from Sourcegraph by going to the compare section from a repository page. You can also see this in action on your [campaigns](#campaigns) to make changes across all the repositories at your organization. 

Take a look at the syntax highlighting on a diff that &lt;WHAT IS THIS DIFF ABOUT> in the Sourcegraph repository.

## Easier to read queries in plain text mode

> TODO: video

Author: @lguychard to add video link to commented out snippet

  </div>
  <p style="text-align: center"><a href="https://youtu.be/{VIDEO_ID}" target="_blank">View on YouTube</a></p>
</p>
-->

In Sourcegraph 3.15, we introduced the AND/OR operators on file contents. In Sourcegraph 3.16, we are removing the experimental flag and making this available to all users.

## Performance improvements for precise code intelligence

![performance-improvements-code-intel](/blog/3-16-ts-to-go-lsif.png "performance-improvements-code-intel")

As of Sourcegraph 3.16, our precise code intelligence backend has been rewritten from TypeScript to Go. This effort is part of a larger effort to aggressively optimize conversion and querying of LSIF data, which will start becoming more apparent in upcoming releases. Even with this initial improvement, we’re seeing improvements on precise code intelligence query performance. The following graph shows the five-run average duration of our [integration test](https://github.com/sourcegraph/sourcegraph/tree/5f51043ad2130a1acdcfca8b969f907cd03a220d/internal/cmd/precise-code-intel-test) before and after the rewrite.

## Discover useful scopes with `repogroup` autocompletion

> TODO: video

Author: @lguychard to add video

## Check syncing status of repository and user permissions

![repository-permissions](https://user-images.githubusercontent.com/2946214/82042090-dd1eaa00-96db-11ea-9fd7-99e0e82a3d0b.png)

Sourcegraph 3.14 and 3.15 introduced a more performant background permissions syncing solution for Bitbucket Server, GitLab, and GitHub repository permissions. Sourcegraph 3.16 adds new permissions pages to user and repository settings when background permissions syncing is enabled. These pages allow admins to check the last time permissions were synced and manually trigger the syncing process for an immediate refresh to the user or repository’s permissions.

Enable this feature by adding ` "permissions.backgroundSync": {"enabled": true}` in your site configuration settings.

## Understanding user activity of your instance

> ![usage-activity-histogram](/blog/3-16-activity-histogram.png "usage activity histogram")

Author: @ebrodymoore - the image from the google doc was low quality - can you upload a higher resolution version here or send to me in Slack?
